cirq/experiments/random_quantum_circuit_generation.py

```diff
@@ -289,16 +294,68 @@ def _get_active_pairs(graph: nx.Graph, grid_layer: GridInteractionLayer):
 
 @dataclasses.dataclass(frozen=True)
 class CircuitLibraryCombination:
-    """For a given layer (which defines a set of pairs of qubits), `combinations` is a 2d array
+    """For a given layer (specifically, a set of pairs of qubits), `combinations` is a 2d array
     of shape (n_combinations, len(pairs)) where each row represents a combination (with replacement)
     of two-qubit circuits. The actual values are indices into a list of library circuits.
 
-    `layer` is used for record-keeping but is otherwise unnecessary. Therefore, it is optional.
+    `layer` is used for record-keeping. This is the GridInteractionLayer if using
+    `get_random_combinations_for_device`, the Moment if using
+    `get_random_combinations_for_layer_circuit` and ommitted if using
+    `get_random_combinations_for_pairs`.
     """
 
-    layer: Optional[GridInteractionLayer]
+    layer: Optional[Any]
     combinations: np.array
-    pairs: List[Tuple['cirq.Qid', 'cirq.Qid']]
+    pairs: List[QidPairT]
+
+
+def _get_random_combinations(
+    n_library_circuits: int,
+    n_combinations: int,
+    *,
+    pair_gen: Generator[Tuple[List[QidPairT], Any], None, None],
+    random_state: 'cirq.RANDOM_STATE_OR_SEED_LIKE' = None,
+) -> List[CircuitLibraryCombination]:
+    """For qubit pairs, prepare a set of combinations to efficiently sample
+    parallel two-qubit XEB circuits.
+
+    This helper function should be called by one of `xx`, `xx` which define
+    appropriate `pair_gen` arguments.
+
+    Args:
+        n_library_circuits: The number of circuits in your library. Likely the value
+            passed to `generate_library_of_2q_circuits`.
+        n_combinations: The number of combinations (with replacement) to generate
+            using the library circuits. Since this function returns a
+            `CircuitLibraryCombination`, the combinations will be represented
+            by indexes between 0 and `n_library_circuits-1` instead of the circuits
+            themselves. The more combinations, the more precise of an estimate for XEB
+            fidelity estimation, but a corresponding increase in the number of circuits
+            you must sample.
+        pair_gen: A generator that yields tuples of (pairs, layer_meta) where pairs is a list
+            of qubit pairs and layer_meta is additional data describing the "layer" assigned
+            to the CircuitLibraryCombination.layer field.
+        random_state: A random-state-like object to seed the random combination generation.
+
+    Returns:
+        A list of `CircuitLibraryCombination`, each corresponding to a layer
```

Comment:
Not quite, you skip the layers with no pairs.

Reply:
Yeah good point. I moved this check to `get_random_combinations_for_device` where it is relevant (for a given device topology sometimes one of the four grid interaction layers doesn't apply, especially for small or narrow devices). I added a note to the return docstring for that function.

The other `get_random_combinations_xxx` functions should have a 1:1 correspondence. The user shouldn't be passing in empty moments or an empty list of pairs.